cpp/dolfinx/fem/interpolate.h

```diff
@@ -598,8 +629,252 @@ void interpolate(Function<T>& u, const Function<T>& v,
+      // process i
+      std::vector<std::int32_t> sendingOffsets(nProcs, 0);
+      for (int i = 0; i < nProcs; ++i)
+      {
+        for (int j = 0; j < mpi_rank; ++j)
+        {
+          sendingOffsets[i] += allPointsToSend[nProcs * j + i];
+        }
+      }
+
+      // Allocate memory to receive coordinate points from other processes
+      xt::xtensor<double, 2> pointsToReceive(
+          xt::shape({nPointsToReceive / 3,
+                     static_cast<decltype(nPointsToReceive)>(3)}),
+          0);
+      // Open a window for other processes to write into, then sync
+      MPI_Win window;
+      MPI_Win_create(pointsToReceive.data(), sizeof(double) * nPointsToReceive,
+                     sizeof(double), MPI_INFO_NULL, MPI_COMM_WORLD, &window);
+      MPI_Win_fence(0, window);
+
+      // Send data to each process. It is intended that sending to itself is the
+      // same as just copying
+      for (int i = mpi_rank; i < nProcs + mpi_rank; ++i)
+      {
+        const auto dest = i % nProcs;
+        MPI_Put(pointsToSend[dest].data(), pointsToSend[dest].size(),
+                MPI_DOUBLE, dest, sendingOffsets[dest],
+                pointsToSend[dest].size(), MPI_DOUBLE, window);
+      }
+
+      // Sync, then close window
+      MPI_Win_fence(0, window);
+      MPI_Win_free(&window);
+
+      // Each process will now check at which points it can evaluate
+      // the interpolating function, and note that down in evaluationCells
+      std::vector<std::int32_t> evaluationCells(pointsToReceive.shape(0), -1);
+
+      const auto connectivity = v.function_space()->mesh()->geometry().dofmap();
+
+      // This BBT is useful for fast lookup of which cell contains a given point
+      dolfinx::geometry::BoundingBoxTree bbt(*v.function_space()->mesh(), tdim,
+                                             0.0001);
+
+      const auto xv = v.function_space()->mesh()->geometry().x();
+
+      // For each point at which the source function needs to be evaluated
+      for (decltype(pointsToReceive.shape(0)) i = 0;
+           i < pointsToReceive.shape(0); ++i)
+      {
+        // Get its coordinates
+        const double xp = pointsToReceive(i, 0);
+        const double yp = pointsToReceive(i, 1);
+        const double zp = pointsToReceive(i, 2);
+
+        // Check if any cells actually contain that point
+        const auto collidingCells = dolfinx::geometry::compute_colliding_cells(
+            *v.function_space()->mesh(),
+            dolfinx::geometry::compute_collisions(
+                bbt, xt::xtensor<double, 2>({{xp, yp, zp}})),
+            xt::xtensor<double, 2>({{xp, yp, zp}}));
+        const std::vector<std::int32_t> intersectingCells(
+            collidingCells.links(0).cbegin(), collidingCells.links(0).cend());
+
+        // If there is any -- there should be at most one -- note it down
+        if (not intersectingCells.empty())
+        {
+          evaluationCells[i] = intersectingCells[0];
+        }
+      }
```

Comment:
Similar comment as above. The compute_colliding cells and compute collisions works on sets of points, and should not be looped over. There is alot unecessary copying here as a result.

Reply:
I think when I first wrote this piece of code those functions only accepted a single point. In any case, I applied the change you suggested.